test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -1622,18 +1625,50 @@ private synchronized void stopNodesAndClient(NodeAndClient nodeAndClient) throws
     }
 
     private synchronized void stopNodesAndClients(Collection<NodeAndClient> nodeAndClients) throws IOException {
+        final Set<String> withdrawnNodeIds = new HashSet<>();
+
         if (autoManageMinMasterNodes && nodeAndClients.size() > 0) {
-            int masters = (int)nodeAndClients.stream().filter(NodeAndClient::isMasterEligible).count();
-            if (masters > 0) {
-                updateMinMasterNodes(getMasterNodesCount() - masters);
+
+            final long currentMasters = nodes.values().stream().filter(NodeAndClient::isMasterEligible).count();
+            final long stoppingMasters = nodeAndClients.stream().filter(NodeAndClient::isMasterEligible).count();
+
+            assert stoppingMasters <= currentMasters : currentMasters + " < " + stoppingMasters;
+            if ((stoppingMasters != currentMasters && currentMasters <= stoppingMasters * 2) || rarely()) {
+                // stopping fewer than _all_ of the master nodes, but at least half of them, requires their votes to be withdrawn first
```

Comment:
I'm wondering if we also need to wait on subsequent shutdowns for the cluster to be optimally reconfigured before returning from this method.
Assume a five node cluster, where you shut down three nodes, one after another. If the cluster has not finished reconfiguring before shutting down the third node, you're in trouble, yet this is nowhere enforced here afaics.

Reply:
Yes, we're assuming that these reconfigurations happen fast enough to avoid this. The `currentMasters <= stoppingMasters * 2` is a proxy for the actual condition to check, which is that the set of remaining nodes still forms a quorum. It's tricky because "forms a quorum" is a fluid statement thanks to reconfiguration.

This could crop up in production too, I think?

Reply:
Ok we will add a health check for verifying that the configuration is optimal, but for now we always withdraw votes: efbaf1b01f8b975b301689708885cd4985e5ca00.